### PR TITLE
Handle reset and a few other fixes

### LINKED ISSUE #9: ioctl(USB_RAW_IOCTL_EP_READ): Cannot send after transport endpoint shutdown when changing altinterface
[log](https://gist.github.com/RandomOnlineName/5108f379b04a59b501c2f3becffea024)
[kmsg](https://gist.github.com/RandomOnlineName/4bc3095e9da44bf402871527137be6bc)

This issue when passing through an xbox controller and with my raspberry pi 4B
When trying to run the program the program hangs for 5-10 second before crashing an printing the error
When switching interfaces the program tries to terminate endpoints but one of them hangs when reading and from what I can tell that causes the crash
The proxy works perfectly using dummy raw gadget and also another device (usb drive) which changes interfaces
A wireshark file as a result of proxying the controller is [here](https://drive.google.com/file/d/1JT5Ksq7Iak__DaNpQbEKZpmWljQL8oyy/view?usp=sharing)

### PR DESCRIPTION
**Pull request by xairy**
Add handling of the `USB_RAW_EVENT_RESET` Raw Gadget event and fix a few other issues.

See the commit descriptions for details.

Fixes #9.

### CODE REVIEW

**Comment by AristoChen** [comment]
Hi @xairy , 

Thanks for the PR! and sorry that I am about to go on a business trip abroad, so I can only review after I am back

File: device-libusb.cpp [hunk 1]
```diff
@@ -122,7 +122,7 @@ int connect_device(int vendor_id, int product_id) {
 		return result;
 	}
 
-	result = libusb_set_auto_detach_kernel_driver(dev_handle, 1);
+	result = libusb_set_auto_detach_kernel_driver(dev_handle, 0);
```
Comment (AristoChen) [inline device-libusb.cpp:125]: May I know the reason to set the value to 0? does it fix any issues? Thanks!

Comment (xairy) [reply device-libusb.cpp:125]: Yes, it prevents the kernel from attaching a driver to the interface when the proxy calls `release_interface` during a device reset or an altsetting change. I think this was the original intention of the `libusb_set_auto_detach_kernel_driver` call, but it used a wrong argument by mistake.

File: proxy.cpp [hunk 2]
```diff
@@ -331,6 +359,37 @@ void ep0_loop(int fd) {
...
+		if (event.inner.type == USB_RAW_EVENT_RESET || event.inner.type == USB_RAW_EVENT_DISCONNECT) {
+			printf("Resetting device\n");
+			// Normally, we would need to stop endpoint threads first and only then
+			// reset the device. However, libusb does not allow interrupting queued
+			// requests submitted via sync I/O. Thus, we reset the proxied device to
+			// force libusb to interrupt the requests and allow the endpoint threads
+			// to exit on please_stop_eps checks.
+			if (set_configuration_done_once)
+				please_stop_eps = true;
+			reset_device();
```
Comment (imatespl) [inline proxy.cpp:373]: hi@xairy, when reset_device, usb hotplug event will be trigger, the usb-proxy  process will restart by  https://github.com/AristoChen/usb-proxy/blob/c681f1238371fae331482f7d300a4652586d0dd3/device-libusb.cpp#L19

**Comment by AristoChen** [comment]
Hi @xairy and @imatespl , 

Let's bring the [conversation](https://github.com/AristoChen/usb-proxy/commit/514403d3c530825bc98958425587bd8527304ed5) to here if you don't mind, I noticed that I won't get any notifictation if you have new comments there

thanks!

**Comment by AristoChen** [comment]
Looks like there is no further discussion, so I will merge this PR soon, and also I will open a new issue to make sure we will have a fix for the issue discussed [here](https://github.com/AristoChen/usb-proxy/commit/514403d3c530825bc98958425587bd8527304ed5#r133307460)

**Review by AristoChen — APPROVED** [review]

**Comment by xairy** [comment]
> Looks like there is no further discussion, so I will merge this PR soon, and also I will open a new issue to make sure we will have a fix for the issue discussed here

Sorry for not responding. I was meaning to get back to this, but I've been busy with other things. Thank you for merging!

**Comment by AristoChen** [comment]
> Sorry for not responding. I was meaning to get back to this, but I've been busy with other things. Thank you for merging!

No worry, I know people get very busy from time to time, even I am not available all the time maintaining this repo, and thanks for the contributions! we can continue the discussion in #20 if you have any thought. thanks again!

**Pull request merged**